### PR TITLE
run-on-vm: add support for running checkin tests on a vm

### PR DESCRIPTION
**Pull request by jeffkaufman**
Split the 'stand up a vm / delete it when done' portion of build_on_vm.sh
off into a new script, run_on_vm.sh, and add checkin_on_vm.sh that uses it
to run checkin tests.

### CODE REVIEW

File: install/run_on_vm.sh [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+# Display an error including the machine name if we die in the script below.
+trap '[ $? -ne 0 ] && echo -e "\nScript failed on $machine_name"' EXIT
+
+echo "Checking whether $machine_name is up; you may see some errors..."
+while ! gcloud compute ssh "$machine_name" -- true; do
+  echo "Waiting for $machine_name to come up..."
+done
+
+machine_ready
```
Comment (hillsp) [inline install/run_on_vm.sh:83]: Hmm, given the way you've done this, what do you think about instead just using the trap on line 76 to delete the machine, then you don't have the awkward callback into machine_ready.

File: install/build_on_vm.sh [hunk 2]
```diff
@@ -99,28 +94,13 @@ if [ -z "$machine_name" ]; then
...
-if [ -z "$instances" ] || ! $use_existing_machine; then
-  gcloud compute instances create "$machine_name" \
-    --image-family="$image_family" --image-project="$image_project"
-fi
-
 mkdir -p ~/release
 
-# Display an error including the machine name if we die in the script below.
-trap '[ $? -ne 0 ] && echo -e "\nBuild failed on $machine_name"' EXIT
+remaining_arguments=( "$@" "" )
```
Comment (hillsp) [inline install/build_on_vm.sh:100]: Why the trailing "" ? That doesn't look right

File: install/build_on_vm.sh [hunk 3]
```diff
@@ -130,11 +110,10 @@ gcloud compute ssh "$machine_name" -- bash << EOF
   fi
   git clone -b "$branch" https://github.com/pagespeed/mod_pagespeed.git
   cd mod_pagespeed
-  install/build_release.sh $@
+  install/build_release.sh ${remaining_arguments[@]}
```
Comment (hillsp) [inline install/build_on_vm.sh]: Oh, was the trailing quotes on remaining_arguments to deal with set -u here?

Maks' solution for this was to test the size of the array first, a la:
if [ "${#remaining_arguments[@]}" -gt 0 ]; then
  install/build_release.sh ${remaining_arguments[@]}
else
  install/build_release.sh
fi

The fancier approach I used in centos/install_required_packages.sh is:
install/build_release.sh ${remaining_arguments[@]+"${remaining_arguments[@]}"}

(from http://stackoverflow.com/questions/7577052/bash-empty-array-expansion-with-set-u)

File: devel/checkin_on_vm.sh [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+  sanitized_branch="$(tr _ - <<< "$branch" | tr -d .)"
+  machine_name="${USER}-checkin-${sanitized_branch}"
+fi
+
+remaining_arguments=( "$@" "" )
+
+# Hook for run_on_vm.sh to call.
+function machine_ready() {
+  gcloud compute ssh "$machine_name" -- bash << EOF
+  set -e
```
Comment (hillsp) [inline devel/checkin_on_vm.sh:76]: Do you want a set -u here?

File: devel/checkin_on_vm.sh [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  set -x
+  sudo apt-get -y update
+  sudo apt-get -y upgrade
+  sudo apt-get -y install git
+  if ! [ -d mod_pagespeed ]; then
```
Comment (hillsp) [inline devel/checkin_on_vm.sh:83]: Why not just git clone --recursive ?

File: devel/checkin_on_vm.sh [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    git clone -b "$branch" https://github.com/pagespeed/mod_pagespeed.git
+  fi
+  cd mod_pagespeed
+  git submodule update --init --recursive
+  if ! [ -d ~/apache2 ]; then
+    install/build_development_apache.sh 2.2 prefork
+  fi
+  devel/checkin ${remaining_arguments[@]}
```
Comment (hillsp) [inline devel/checkin_on_vm.sh]: Quote remaining arguments

**Review by hillsp — CHANGES_REQUESTED** [review]

File: install/build_on_vm.sh (hunk 2 above)
Comment (jeffkaufman) [reply install/build_on_vm.sh:100]: added comment 9488d44

File: install/build_on_vm.sh (hunk 3 above)
Comment (jeffkaufman) [reply install/build_on_vm.sh]: Just doing `( "$@" "" )` seems much tidier to me?

File: devel/checkin_on_vm.sh (hunk 6 above)
Comment (jeffkaufman) [reply devel/checkin_on_vm.sh]: done (and in build_on_vm.sh): bbdf299

File: devel/checkin_on_vm.sh (hunk 5 above)
Comment (jeffkaufman) [reply devel/checkin_on_vm.sh:83]: done: d8844a0

File: devel/checkin_on_vm.sh (hunk 4 above)
Comment (jeffkaufman) [reply devel/checkin_on_vm.sh:76]: done: 0176598

File: install/run_on_vm.sh (hunk 1 above)
Comment (jeffkaufman) [reply install/run_on_vm.sh:83]: We don't want to delete the machine if the thing we run on it fails, because someone may want to ssh in and look at it.

**Comment by jeffkaufman** [comment]
Ready for another round of review

File: install/run_on_vm.sh (hunk 1 above)
Comment (hillsp) [reply install/run_on_vm.sh:83]: Take a look at how the trap works, you can test $? and in fact I already do.

File: install/build_on_vm.sh (hunk 3 above)
Comment (hillsp) [reply install/build_on_vm.sh]: That's going to insert an empty last argument, which I don't think you want?

**Pull request closed without merging**